Restore color rendering following Electron 2.0 update

jasonrudolph opened the pull request:
Fixes #17356

## Description of the Change

### Background

In https://github.com/atom/atom/issues/17356 and https://github.com/electron/electron/issues/10732, users reported an unwelcome change to Atom's color rendering following the recent update to Electron 2.0.

This [doc from the Chrome team](https://docs.google.com/document/d/1jMokB_OBkZVELu22li8vnHxAUoL1eGnLedP-1Gttv40/edit) (shared in https://github.com/electron/electron/issues/10732#issuecomment-390288101) explains that these color differences are the result of a change in Chromium, and that the change is intentional:

> Historically, Chrome has been inconsistent in how it handles colors, and would make no effort to ensure that a web designer’s color selection look the same on all monitors. As of Chrome 61, we respect monitors’ color profiles, so all web pages should appear the same on all devices.

However, the document goes on to acknowledge that this change has received a fair bit of pushback:

> Many users have filed bugs that Chrome looks different now. Most of these users have been on Linux or Windows, have set a color profile for their monitor (or have had one set for them), but do not want Chrome to respect this color profile. ... The users’ system settings specify how colors should be interpreted on their monitor, but there exist users who are used to the “wrong” colors that they had before.

Chromium provides a flag that allows you to force a particular color profile:

> The [“Force color profile” option] tells Chrome to ignore the operating system settings and use some other value instead. Setting this to “sRGB” will make Chrome behave as it did before on Linux and Windows (for 99% of content).

### Approach

This pull request allows Atom users to opt in to using the `force-color-profile` chromium flag.

#### Usage

To restore Atom's previous color behavior:

1. Open Atom's settings UI (i.e., `Settings View: Open` in the command palette)
2. Click the "Core" tab, find the "Color Profile" option, and choose "Use sRGB color profile"
3. Restart Atom

<img width="994" alt="color-profile-settings" src="https://user-images.githubusercontent.com/2988/41253300-e56ce36c-6d8d-11e8-8871-6ccb0bd5c6b7.png">

## Alternate Designs

- We could _always_ force the color profile to `srgb`.

  However, as described in the [doc from the Chrome team](https://docs.google.com/document/d/1jMokB_OBkZVELu22li8vnHxAUoL1eGnLedP-1Gttv40/edit), this would essentially "lock all users forever into 1996-era technology," and we'd prefer to avoid that.

- We could do nothing, but some users want Atom to behave like *most* other apps by ignoring the user's custom color profile. https://github.com/atom/atom/issues/17356#issuecomment-391922655, https://github.com/atom/atom/pull/17380#issuecomment-393016583

## Benefits

Provides an option to instruct Atom to ignore the user's custom color profile, thus causing Atom 1.28's color rendering to match the rendering used in Atom 1.27 and earlier (i.e., the color rendering as it existed prior to the Electron 2.0 upgrade in #17273).

## Possible Drawbacks

Adding more code to the startup process means there are more things that can go wrong in the startup process. If we encounter problems reading and/or parsing the config file, it may cause Atom to crash at startup in ways that are difficult to debug.

## Verification Process

Using Ubuntu 18.04, build Atom from this branch (which uses Electron 2.0.1) and verify that:

- [x] With a custom color profile configured:
    - [x] When no config file exists, Atom launches successfully and respects the user's custom color profile
    - [x] When the `colorProfile` option is not present in the config file, Atom launches successfully and respects the user's custom color profile
    - [x] When the `colorProfile` option is present in the config file and has a value of `"srgb"`, Atom ignores the user's custom color profile, and Atom's color rendering matches the color rendering seen in Atom 1.27.1
- [x] Without a custom color profile configured, when the `colorProfile` option is present in the config file and has a value of `"srgb"`, Atom launches successfully and Atom's color rendering matches the color rendering seen in Atom 1.27.1
- [x] Changing an existing setting (e.g., font size) is still possible via Atom's settings UI

---

/fyi @codebytere

jasonrudolph committed:
Fix #17356

Refs:
https://github.com/electron/electron/issues/10732#issuecomment-390321086

jasonrudolph committed:
Merge master into fix-color-problems-using-electron-2.0

Arcanemagus commented:
> Maybe we could alter Atom's CSS in some way to cope with Chromium's updated color rendering behavior?

The problem isn't with the colors in the CSS, but that the colors are now accurate, where before they were oversaturated due to the incorrect mapping being done before. When you are used to the incorrect, oversaturated, colors when they are suddenly now rendered correctly it usually looks "washed out".

> there may be some Atom users that feel that the color "regression" reported in #17356 is actually an enhancement, and they may consider it to be a bug to restore the old color rendering.

I _definitely_ fall into that category, you should always be rendering the correct colors 😉.

jasonrudolph commented:
> ### Verification Process
>
> Build Atom from this branch (which uses Electron 2.0.1) and verify that this branch's color rendering matches the color rendering seen in Atom 1.27.1 (which uses Electron > 1.7.15):
>
> - [x] Linux
>     - [x] With custom display profile configured
>     - [x] Without custom display profile configured

### With custom display profile configured

#### Atom 1.27.1

<img width="1374" alt="2018-05-21 at 13 26 16 pm ubuntu 1 27 with custom color profile" src="https://user-images.githubusercontent.com/2988/40321195-41fa3028-5cfc-11e8-9aea-d94b8ef98f09.png">

#### Atom built from this branch

<img width="1373" alt="2018-05-21 at 13 29 16 pm ubuntu 1 29 with custom color profile" src="https://user-images.githubusercontent.com/2988/40321196-420813fa-5cfc-11e8-826c-c93c0748009d.png">

### Without custom display profile configured

#### Atom 1.27.1

<img width="1372" alt="2018-05-21 at 13 24 50 pm ubuntu 1 27 with default color profile" src="https://user-images.githubusercontent.com/2988/40321194-41e142b6-5cfc-11e8-9519-d2e802ed752f.png">

#### Atom built from this branch

<img width="1372" alt="2018-05-21 at 13 30 16 pm ubuntu 1 29 with default color profile" src="https://user-images.githubusercontent.com/2988/40321197-421ab1e0-5cfc-11e8-83e4-d4fcb88ad9a5.png">

jasonrudolph commented:
> ### Verification Process
>
> Build Atom from this branch (which uses Electron 2.0.1) and verify that this branch's color rendering matches the color rendering seen in Atom 1.27.1 (which uses Electron > 1.7.15):
>
> - [ ] macOS
>     - [x] ✅ Without custom display profile configured
>     - [ ] With custom display profile configured 

### Without custom display profile configured ✅

Using the display's default profile ...

<img width="1440" alt="2018-05-21 at 13 07 29 pm macos with default color profile" src="https://user-images.githubusercontent.com/2988/40321621-9fa580f0-5cfd-11e8-8524-6ff386f757c9.png">

I see the following results ...
<img width="780" alt="screen shot 2018-05-21 at 13 50 50 pm" src="https://user-images.githubusercontent.com/2988/40321764-21817c00-5cfe-11e8-916c-46635b23926d.png">

### With custom display profile configured ❌

Using this profile ...

<img width="780" alt="screen shot 2018-05-21 at 13 50 54 pm" src="https://user-images.githubusercontent.com/2988/40321765-219aeec4-5cfe-11e8-93a9-5402afc5121d.png">

I see the following results ...

<img width="1440" alt="2018-05-21 at 13 07 29 pm macos with custom color profile" src="https://user-images.githubusercontent.com/2988/40321620-9f980eb6-5cfd-11e8-9434-2c9869162a84.png">

jasonrudolph commented:
> ### Verification Process
>
> Build Atom from this branch (which uses Electron 2.0.1) and verify that this branch's color rendering matches the color rendering seen in Atom 1.27.1 (which uses Electron > 1.7.15):
>
> - [x] Windows
>     - [x] With custom display profile configured
>     - [x] Without custom display profile configured

### Without custom display profile configured ✅

Using the display's default profile, I see the following results:

<img width="1440" alt="default-display-profile" src="https://user-images.githubusercontent.com/2988/40328548-c2fc0ab2-5d14-11e8-99e9-cd4e15bf2525.png">

### Without custom display profile configured ✅

Using a [custom display profile (Adobe RGB 1998)](https://user-images.githubusercontent.com/2988/40328549-c311ec10-5d14-11e8-8843-6a4cd1c2cb25.png), I see the following results:

<img width="1440" alt="custom-display-profile" src="https://user-images.githubusercontent.com/2988/40328547-c2e9dcf2-5d14-11e8-8aac-bd35237b9d85.png">

simurai commented:
@jasonrudolph Thanks for taking all the screenshots. 👌 

I guess the "safe thing to do ™️" would be to merge this PR to avoid change. 

But the "right thing to do ™️" _would_ probably be to use the system's color profile. On macOS and Windows the differences seem to be noticeable, but it doesn't make Atom "unusable". 

Just on Linux, that example with the custom profile:

<img width="1379" alt="screen shot 2018-05-21 at 14 57 26 pm" src="https://user-images.githubusercontent.com/2988/40324852-d1ec16b4-5d07-11e8-9e22-78b820a529f8.png">

That's quite the drastic change and makes it hard to use. But then I wonder, people that actually use that color profile, they _should_ be used to not a lot of contrast with dark backgrounds, so they are probably using a light theme instead?

So with all that in mind, I lean slightly towards **not** merging this PR, even tough some people might get upset. Because you can also look at it from this angle: Atom finally respects the system's color profile that the user can tweak and doesn't force you into using `sRGB`, like it before.

simurai commented:
Made a test similar to the one in [discuss](https://discuss.atom.io/t/any-way-to-change-color-rendering-in-atom-1-28-0-beta-electron-2-0-0/55234/5?u=simurai).

The CSS value of the background color is: `rgb(42, 42, 60)`:

![screen shot 2018-05-23 at 9 30 23 pm](https://user-images.githubusercontent.com/378023/40424863-ae5a738c-5ed1-11e8-860c-fdeaed951f3e.png)

- When using the default profile for my monitor `Thunderbolt Display`, the value with the color picker is:
  - Atom 1.28: `rgb(42, 43, 61)`
  - This branch: `rgb(42, 42, 60)`
- When switching to the `sRGB` profile it's exactly like authored:
  - Atom 1.28: `rgb(42, 42, 60)`
  - This branch: `rgb(42, 42, 60)`

![colors](https://user-images.githubusercontent.com/378023/40425858-91ff08f8-5ed4-11e8-96be-f3da068f361f.gif)

So by using the `sRGB` color profile in your system settings, you might can change it back "how it used to be".

jasonrudolph commented:
> ### Possible Drawbacks
>
> According to the [Google doc](https://docs.google.com/document/d/1jMokB_OBkZVELu22li8vnHxAUoL1eGnLedP-1Gttv40/edit) shared in https://github.com/electron/electron/issues/10732#issuecomment-390288101, the updated colors are the "correct" colors, and users are surprised by the updated colors because they "are used to the 'wrong' colors that they had before." Following that line of thought, there may be some Atom users that feel that the color "regression" reported in #17356 is actually an enhancement, and they may consider it to be a bug to restore the old color rendering.

The more I research this topic, the more I'm in favor of closing this pull request. Given the context in the [Google doc from the Chrome dev team](https://docs.google.com/document/d/1jMokB_OBkZVELu22li8vnHxAUoL1eGnLedP-1Gttv40/edit) and the resources that the doc links to, Atom 1.28's color rendering seems to be a move in the right direction (as opposed to being a regression).

https://github.com/atom/atom/issues/17356#issuecomment-391450897 describes my current understanding of the situation, and I welcome feedback. :bow:

daviwil commented:
I'm also in favor of closing this PR and keeping things as they are.  Thanks for the extremely thorough investigation @jasonrudolph! :bowing_man:

simurai commented:
Having thought about it some more.. I'm in favor of this alternative:

> We could potentially make this behavior configurable, either via a command-line switch or a preference specified by the user's local Atom configuration. In other words, the user could choose whether that want to force Atom to use a particular color profile.

Add a config option: `[ ] Use the sRGB color profile`. Disabled by default.

It **is** yet another checkbox to maintain. :see_no_evil: But when Atom 1.28 goes to stable, we would have to tell people that experience this issue the following:

- **Without** config: "Please change your custom color profile"
- **With** config: "Please consider changing your custom color profile or enable the sRGB color profile in Atom's settings"

I just feel that "forcing" people to change their custom color profile is a bit too much to ask.

jasonrudolph committed:
Read user's custom Electron switches from config file

jasonrudolph committed:
Look for .electron-switches in ATOM_HOME

jasonrudolph commented:
I've updated the implementation to allow users to **opt in** to using the force-color-profile chromium flag. I've updated the pull request body to reflect this approach.

@daviwil @maxbrunsfeld: Would you mind reviewing this pull request? I'd love to get your feedback on the design (as described in the pull request body) and on the implementation.

jasonrudolph committed:
Use Atom's standard config file to set custom color profile

jasonrudolph commented:
I've updated the implementation to use Atom's standard config file (i.e., `config.cson` or `config.json` in your Atom home directory) to store this config option. I've updated the pull request body accordingly.

jasonrudolph committed:
:shirt:

jasonrudolph committed:
☠☕ Decaffienate ScopeDescriptor

Oddly, I expect this to resolve the CI failure seen in 
https://travis-ci.org/atom/atom/builds/386819124#L1200.

jasonrudolph committed:
Gracefully handle missing config file

https://circleci.com/gh/atom/atom/7630

jasonrudolph committed:
Support configuring color profile via settings UI

jasonrudolph commented:
Let's look at the behavior before and after these changes.

### Before

https://github.com/atom/atom/issues/17356#issuecomment-391450897 demonstrated the behavior in Atom 1.28.0-beta1. For this test, we observed how the background color of the text editor is rendered when using Atom's [One Dark syntax theme](https://atom.io/themes/one-dark-syntax).
We saw that when applying the Adobe RGB 1998 color profile, Chrome 66 and Atom 1.28.0-beta1 both render `#282c34` as `#2d3037`:

![demo](https://user-images.githubusercontent.com/2988/40439553-4110523a-5e89-11e8-9a26-97313f34d3a2.gif)

### After

With the changes in this pull request, Atom still respects the operating system's color profile **by default**, but the user can optionally force Atom to use an sRGB color profile. We see this in the demo below.

When applying the Adobe RGB 1998 color profile to the operating system, by default Atom respects the color profile and renders `#282c34` as `#2d3037`. When instructing Atom to use an sRGB color profile, Atom renders `#282c34` as `#282c34`.

![demo](https://user-images.githubusercontent.com/2988/40864443-d6d0919c-65c1-11e8-8f4e-f38af28b4b6c.gif)

maxbrunsfeld commented on src/main-process/start.js:
```diff
@@ -39,6 +42,12 @@ module.exports = function start (resourcePath, startTime) {
   atomPaths.setUserData(app)
   setupCompileCache()
 
+  const config = getConfig()
```
It would be *nice* if we could reuse this `Config` and/or `ConfigFile` instance in `AtomApplication`, in order to avoid reading the config file twice on startup (and also to avoid duplicating the code for locating the config file). That said, I don't think it is necessarily important enough to block merging this PR.

maxbrunsfeld reviewed (approved):
This looks great! Thanks for taking the time to address the UX around color profiles so thoroughly.

daviwil commented:
Excellent work @jasonrudolph!  I'm going to quickly test this out on Windows and see if it has the desired effect.

jasonrudolph commented:
I chatted with @daviwil regarding the next steps for this pull request:

- David is going to test on Windows
- If all is well with testing, David will merge this pull request and then ship an Atom 1.28-beta hotfix next week

Note: We acknowledge that Atom doesn't automatically relaunch when you change this new config setting, but that's something that could be addressed in a follow-up pull request. 😅

daviwil commented:
Tested this out on Windows, works great!

With _Color Profile_ set to "Use color profile configured in the operating system" the colors are washed out:

![1 28-os-profile](https://user-images.githubusercontent.com/79405/40868375-1fbcc228-65c1-11e8-9205-3637bd496361.PNG)

Once I change _Color Profile_ to "Use sRGB color profile" and restart Atom, I see this:

![1 28-srgb-profile](https://user-images.githubusercontent.com/79405/40868376-2170f01c-65c1-11e8-8a7d-2c8ad9e2ada9.PNG)

Thumbs up from me!  I'll get this cherry-picked over to `1.28-releases` on Monday and ship a hotfix after that.

simurai reviewed (approved):
> Note: We acknowledge that Atom doesn't automatically relaunch when you change this new config setting, but that's something that could be addressed in a follow-up pull request.

Yeah, would be nice to be consistent with the custom title-bar setting, but not that high priority.

thomasjo commented:
I implemented the restart prompt in [another branch](https://github.com/atom/atom/tree/extended-color-profile-bits). Let me know if you want me to cherry-pick the commit(s) into this branch.

---
1. 9b5862179951cbd84576d6cc89f35f7706c18501
2. c80e70192637d3ad29e81c7bc284a4faeb2b9166
3. c7935c5223b002be5334de640630dcfbb2102816

daviwil commented:
Thanks @thomasjo!  I'll go ahead and merge this PR now and then we can merge your improvement afterward.

atomiks commented:
Thanks for this setting! I actually love the oversaturated/vibrant syntax colors of the `sRGB` color profile, but prefer using the default Color LCD profile on my computer because otherwise images look oversaturated.

atomiks commented:
For some reason this isn't working for me under 1.28.0 on macOS 10.13.4. The OS color profile is "Color LCD" (default).

The `Use sRGB color profile` is the same as the default one, so the colors are desaturated now 😢 .

jasonrudolph commented:
> The `Use sRGB color profile` is the same as the default one, so the colors are desaturated now

@atomiks: Thanks for reporting this. I see that you've opened https://github.com/atom/atom/issues/17564, so let's use that issue to discuss the behavior you're seeing.
